custom_components/tuya_local/goldair_portable_airconditioner.yaml

```diff
@@ -0,0 +1,126 @@
+name: 4-in-1 Portable Air Conditioner
+products:
+  - id: gylvqvz3p94i3zdn
+    name: Goldair Portable Air Conditioner
+primary_entity:
+  entity: climate
+  dps:
+    - id: 1
+      type: boolean
+      name: hvac_mode
+      mapping:
+        - dps_val: false
+          value: "off"
+        - dps_val: true
+          constraint: mode
+          conditions:
+            - dps_val: cold
+              value: cool
+            - dps_val: hot
+              value: heat
+            - dps_val: wet
+              value: dry
+            - dps_val: wind
+              value: fan_only
+            - dps_val: heat
+              value: heat
+
+    - id: 2
+      type: integer
+      name: temperature
+      range:
+        min: 16
+        max: 31
+      unit: C
+    - id: 3
+      type: integer
+      name: current_temperature
+      # 107, 108 listed as F temperatures in report, but not existing
+      # in log, and there is no temperature_unit dp to work with
+    - id: 4
+      type: string
+      name: mode
+      hidden: true
+    - id: 5
+      type: string
+      name: fan_mode
+      mapping:
+        - dps_val: low
+          value: low
+        - dps_val: med
+          value: medium
+        - dps_val: high
+          value: high
+    - id: 15
+      type: string
+      name: swing_1
+    - id: 20
+      type: bitfield
+      name: fault_code
+    - id: 109
+      type: integer
+      name: feature_flags
```

Comment:
It might be useful to decode these. Is there any information in the description of the dp on platform.tuya.com?

Reply:
It should be fixed for a given device. It will be an encoded list of feature flags for features the device supports. Things like the missing temperature_unit dp, ionizer availability, and whether the horizontal swing is available should be indicated here.

If there is any documentation, it will be in the "Query Things Data Model" result.

Reply:
Below is what I got. The feature flags is mostly in Chinese but is it saying that ionize isn't supported and only one swing type is supported?

`"{\"modelId\":\"000001zaic\",\"services\":[{\"actions\":[],\"code\":\"\",\"description\":\"\",\"events\":[],\"name\":\"默认服务\",\"properties\":[{\"abilityId\":1,\"accessMode\":\"rw\",\"code\":\"Power\",\"description\":\"\",\"extensions\":{\"iconName\":\"icon-dp_power\",\"attribute\":\"1\"},\"name\":\"开关\",\"typeSpec\":{\"type\":\"bool\"}},{\"abilityId\":2,\"accessMode\":\"rw\",\"code\":\"temp_set\",\"description\":\"\",\"extensions\":{\"iconName\":\"icon-dp_temp\"},\"name\":\"设置温度\",\"typeSpec\":{\"type\":\"value\",\"max\":31,\"min\":16,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":3,\"accessMode\":\"ro\",\"code\":\"temp_current\",\"description\":\"\",\"extensions\":{\"iconName\":\"icon-dp_sun\"},\"name\":\"当前温度 \",\"typeSpec\":{\"type\":\"value\",\"max\":35,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"℃\"}},{\"abilityId\":4,\"accessMode\":\"rw\",\"code\":\"mode\",\"description\":\"cold：制冷模式；wet：除湿模式；wind：送风模式；hot：制热模式；可根据产品实际功能修改功能属性。\",\"extensions\":{\"iconName\":\"icon-dp_mode\"},\"name\":\"工作模式\",\"typeSpec\":{\"type\":\"enum\",\"range\":[\"cold\",\"wet\",\"wind\",\"hot\"]}},{\"abilityId\":5,\"accessMode\":\"rw\",\"code\":\"windspeed\",\"description\":\"低风\\n中风\\n高风\",\"extensions\":{\"iconName\":\"icon-dp_wind\"},\"name\":\"风速\",\"typeSpec\":{\"type\":\"enum\",\"range\":[\"low\",\"med\",\"high\"]}},{\"abilityId\":11,\"accessMode\":\"rw\",\"code\":\"anion\",\"description\":\"\",\"extensions\":{\"iconName\":\"icon-dp_fresh\"},\"name\":\"负离子\",\"typeSpec\":{\"type\":\"bool\"}},{\"abilityId\":15,\"accessMode\":\"rw\",\"code\":\"windshake\",\"description\":\"开关\",\"extensions\":{\"iconName\":\"icon-dp_shake\"},\"name\":\"竖摆风\",\"typeSpec\":{\"type\":\"enum\",\"range\":[\"off\",\"on\"]}},{\"abilityId\":20,\"accessMode\":\"ro\",\"code\":\"Fault\",\"description\":\"1:室温传感器故障；2:盘管传感器故障；3：水满报警；\\n\",\"extensions\":{\"iconName\":\"icon-dp_warming\",\"scope\":\"fault\"},\"name\":\"故障告警\",\"typeSpec\":{\"type\":\"bitmap\",\"label\":[\"E1\",\"E2\",\"warter_full\"],\"maxlen\":3}},{\"abilityId\":103,\"accessMode\":\"rw\",\"code\":\"Sleeping_mode\",\"description\":\"\",\"name\":\"睡眠\",\"typeSpec\":{\"type\":\"bool\"}},{\"abilityId\":104,\"accessMode\":\"rw\",\"code\":\"TimerOn\",\"description\":\"\",\"name\":\"定时开启\",\"typeSpec\":{\"type\":\"value\",\"max\":1440,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"分钟\"}},{\"abilityId\":105,\"accessMode\":\"rw\",\"code\":\"TimerOff\",\"description\":\"\",\"name\":\"定时关闭\",\"typeSpec\":{\"type\":\"value\",\"max\":1440,\"min\":0,\"scale\":0,\"step\":1,\"unit\":\"分钟\"}},{\"abilityId\":107,\"accessMode\":\"rw\",\"code\":\"temp_set_f\",\"description\":\"\",\"name\":\"设置温度\",\"typeSpec\":{\"type\":\"value\",\"max\":88,\"min\":61,\"scale\":0,\"step\":1,\"unit\":\"℉\"}},{\"abilityId\":108,\"accessMode\":\"rw\",\"code\":\"temp_current_f\",\"description\":\"\",\"name\":\"当前温度 \",\"typeSpec\":{\"type\":\"value\",\"max\":95,\"min\":32,\"scale\":0,\"step\":1,\"unit\":\"℉\"}},{\"abilityId\":109,\"accessMode\":\"ro\",\"code\":\"funcTag\",\"description\":\"第一位：竖摆风 0：无竖摆风 1：竖摆风；\\n第二位：横摆风 0：无横摆风 1：横摆风；\\n第三位：无负离子0、负离子1；\\n第四位：单冷0、冷暖1；\\n第五位：风速二档0、三挡1；\\n第六位：摄氏度0，华氏度1；\",\"extensions\":{\"scope\":\"fault\"},\"name\":\"功能标记\",\"typeSpec\":{\"type\":\"bitmap\",\"label\":[\"0\",\"1\",\"2\",\"3\",\"4\",\"5\"],\"maxlen\":6}},{\"abilityId\":110,\"accessMode\":\"rw\",\"code\":\"windshakeH\",\"description\":\"\",\"name\":\"横摆风\",\"typeSpec\":{\"type\":\"bool\"}}]}]}"`

Reply:
So it looks like:

- [ ] Vertical swing
- [X] Horizontal swing
- [ ] Ionizer
- [X] Heating
- [X] 3 speed fan 
- [ ] Fahrenheit

Reply:
The manual says that the C/F setting can be changed by pressing both temperature buttons, if you can set it into F, then the dp value should change I think to 58. The other fields seem more like hardware limitations so probably won't ever change.

Reply:
You are totally right, when I change it to F then the dps changes to 58. That's super interesting! So is there some way to construct this that when in 26 it uses C related dps and when in 58 it uses F dps?